### ISSUE TITLE
Feature: Global `SkipRule` or Generic `BeforeRule` preprocessor

### ISSUE THREAD

**Issue by anorborg** [bchavez/Bogus#178]
This is really an amazing tool, but would appreciate some sugar-syntax in my use-case to be able to vary in some `null`/unset values.

**What I do today**
```csharp
new Faker<Test>()
     .RuleFor(x => x.Value1, f => {
          // Randomly set this value with a probability of .1 (10% of the time)
          return (f.Random.Bool(0.1f)) ? f.Random.Int(1, 5) : null;
      })
     .RuleFor(x => x.Value2, f => {
          // Randomly set this value with a probability of .1 (10% of the time)
          return (f.Random.Bool(0.1f)) ? f.Internet.Ip(): null;
      })
```

**Suggested Feature Syntax**
```csharp
new Faker<Test>()
     .SkipRule((f, x) => f.Random.Bool(0.1f))
     .RuleFor(x => x.Value1, f => f.Random.Int(1, 5))
     .RuleFor(x => x.Value2, f => f.Internet.Ip());
```

I'm not sure if there is a better/current way to accomplish something similar, but generating a dataset with the occasional null value is pretty useful for my in testing scenarios. Again, great tool!

**Comment by bchavez** [comment]
Hi @anorborg,

Thank you for the complements! I'm really glad you like **Bogus**! It always makes me happy to hear from people using **Bogus**. :sunglasses: 

To answer your question:
**Bogus** has a nifty `.OrNull()` helper extension to generate null values randomly; however, the feature is nestled away in an extension namespace called `Bogus.Extensions`. Below is an example of how to use the `.OrNull()` extension to generate randomly null values:

```csharp
using Bogus.Extensions;

void Main()
{
   var fakeTests = new Faker<Test>()
      .RuleFor( x => x.Value1, f => f.Random.Int(1, 5).OrNull(f) )
      .RuleFor( x => x.Value2, f => f.Internet.Ip().OrNull(f) );
      
   fakeTests.Generate(10).Dump();
}

public class Test
{
   public int? Value1 {get; set;}
   public string Value2 {get;set;}
}
```
![linqpad_2187](https://user-images.githubusercontent.com/478118/46066253-14b49900-c129-11e8-9ddd-29ae07a2278b.png)

Currently, `.OrNull()` doesn't support a weighted probability of generated nulls; but that can be easily fixed by [copying the current implementation](https://github.com/bchavez/Bogus/blob/0c90ed620d06b327c12ca78323fc2397086c8c10/Source/Bogus/Extensions/ExtensionsForFakerT.cs#L25-L32) and making your own extension method like this:

```csharp
void Main()
{
   var fakeTests = new Faker<Test>()
      .RuleFor( x => x.Value1, f => f.Random.Int(1, 5).OrNull(f, 0.9f) )
      .RuleFor( x => x.Value2, f => f.Internet.Ip().OrNull(f, 0.1f) );
      
   fakeTests.Generate(10).Dump();
}

public static class MyBogusExtensions
{
   public static object OrNull(this object value, Faker f, float nullWeight)
   {
      return f.Random.Float() < nullWeight ? null : value;
   }
}
```
![linqpad_2188](https://user-images.githubusercontent.com/478118/46067755-732f4680-c12c-11e8-8181-d8ca333725fe.png)

The code above is quite clear, elegant, terse, and very readable IMHO. Perhaps, it might be a good idea to make this change official and allow the user to optionally specify the weight of nullability. :smiley_cat: 

Hope that helps!

Thanks,
Brian

:hourglass_flowing_sand: :mag: ***["But I still haven't found what I'm for..."](https://www.youtube.com/watch?v=wdCJRybAtso)***

**Comment by anorborg** [comment]
Thanks! A small suggestion: use `default` instead of `null` which works for null but value-types as well:

```csharp
public static object OrDefault(this object value, Faker f, double defaultValueWeight = 0.5f)
{
    if (defaultValueWeight > 1 || defaultValueWeight < 0) throw new ArgumentOutOfRangeException(nameof(defaultValueWeight), $".{nameof(OrDefault)}() {nameof(defaultValueWeight)} of '{defaultValueWeight}' must be between 1.0f and 0.0f. ");
        return f.Random.Float() > defaultValueWeight ? value : default;
   }
}
```

Although you can define _zero_ in your range in most, if not all cases, since you've added the weight, its a nice option to only select that value with the same probability as null.

**Comment by bchavez** [comment]
Hi @anorborg,

That's a great idea! Thank you for the suggestion! I've released [**Bogus** `v24.1.0`](https://www.nuget.org/packages/Bogus/24.1.0) with the latest suggestion and changes!

Hope that helps!

Thanks,
Brian

:walking: :sunglasses: ***["So don't delay... act now supplies are running out..."](https://www.youtube.com/watch?v=LQj--Kjn0z8)***
